javalib/src/main/scala/java/util/RedBlackTree.scala

```diff
@@ -0,0 +1,957 @@
+      y = x
+      cmp = compare(key, x.key)
+      x = if (cmp < 0) x.left else x.right
+    }
+
+    if (cmp == 0) {
+      val oldValue = y.value
+      y.value = value
+      oldValue
+    } else {
+      val z = Node.leaf(key, value, red = true, y)
+
+      if (y eq null) {
+        /* Here, the key was never compared to anything. Compare it with itself
+         * so that we eagerly cause the comparator to throw if it cannot handle
+         * the key at all, before we put the key in the map.
+         */
+        compare(key, key)
+        tree.root = z
+      } else if (cmp < 0) {
+        y.left = z
+      } else {
+        y.right = z
+      }
+
+      fixAfterInsert(tree, z)
+      tree.size += 1
+
+      null.asInstanceOf[B]
+    }
+  }
+
+  private[this] def fixAfterInsert[A, B](tree: Tree[A, B],
+      node: Node[A, B]): Unit = {
+    var z = node
+    while (isRed(z.parent)) {
+      if (z.parent eq z.parent.parent.left) {
+        val y = z.parent.parent.right
+        if (isRed(y)) {
+          z.parent.red = false
+          y.red = false
+          z.parent.parent.red = true
+          z = z.parent.parent
+        } else {
+          if (z eq z.parent.right) {
+            z = z.parent
+            rotateLeft(tree, z)
+          }
+          z.parent.red = false
+          z.parent.parent.red = true
+          rotateRight(tree, z.parent.parent)
+        }
+      } else { // symmetric cases
+        val y = z.parent.parent.left
+        if (isRed(y)) {
+          z.parent.red = false
+          y.red = false
+          z.parent.parent.red = true
+          z = z.parent.parent
+        } else {
+          if (z eq z.parent.left) {
+            z = z.parent
+            rotateRight(tree, z)
+          }
+          z.parent.red = false
+          z.parent.parent.red = true
+          rotateLeft(tree, z.parent.parent)
+        }
+      }
+    }
+    tree.root.red = false
+  }
+
+  // ---- deletion ----
+
+  def delete[A, B](tree: Tree[A, B], key: Any)(
+      implicit comp: Comparator[_ >: A]): B = {
+    val z = getNode(tree.root, key)
+    if (z eq null) {
+      null.asInstanceOf[B]
+    } else {
+      deleteNode(tree, z)
+      z.value
+    }
+  }
+
+  def deleteNode[A, B](tree: Tree[A, B], z: Node[A, B]): Unit = {
+    var y = z
+    var yIsRed = y.red
+    var x: Node[A, B] = null
+    var xParent: Node[A, B] = null
+
+    if (z.left eq null) {
+      x = z.right
+      transplant(tree, z, z.right)
+      xParent = z.parent
+    } else if (z.right eq null) {
+      x = z.left
+      transplant(tree, z, z.left)
+      xParent = z.parent
+    } else {
+      y = minNodeNonNull(z.right)
+      yIsRed = y.red
+      x = y.right
+
+      if (y.parent eq z) {
+        xParent = y
+      } else {
+        xParent = y.parent
+        transplant(tree, y, y.right)
+        y.right = z.right
+        y.right.parent = y
+      }
+      transplant(tree, z, y)
+      y.left = z.left
+      y.left.parent = y
+      y.red = z.red
```

Comment:
I have abandoned reviewing this, due to the lack of reasonable names. Consider Please use semantic names like, for example: `replacement` (node we put in place of the node to delete), `orphan` (node that has just lost its parent because it was moved).